web/packages/teleport/src/Roles/RoleList/RoleList.tsx

```diff
@@ -20,60 +20,76 @@ import React from 'react';
 import Table, { Cell } from 'design/DataTable';
 import { MenuButton, MenuItem } from 'shared/components/MenuAction';
 
-import { State as ResourceState } from 'teleport/components/useResources';
+import { SearchPanel } from 'shared/components/Search';
 
-import { State as RolesState } from '../useRoles';
+import { SeversidePagination } from 'teleport/components/hooks/useServersidePagination';
+import { RoleResource } from 'teleport/services/resources';
 
-export default function RoleList({
-  items = [],
-  pageSize = 20,
+export function RoleList({
   onEdit,
   onDelete,
-}: Props) {
+  onSearchChange,
+  search,
+  serversidePagination,
+}: {
+  onEdit(id: string): void;
+  onDelete(id: string): void;
+  onSearchChange(search: string): void;
+  search: string;
+  serversidePagination: SeversidePagination<RoleResource>;
+}) {
   return (
     <Table
-      data={items}
+      data={serversidePagination.fetchedData.agents}
+      fetching={{
+        fetchStatus:
+          serversidePagination.attempt.status === 'processing' ? 'loading' : '',
+        onFetchNext: serversidePagination.fetchNext,
+        onFetchPrev: serversidePagination.fetchPrev,
+      }}
+      serversideProps={{
+        sort: undefined,
+        setSort: () => undefined,
+        serversideSearchPanel: (
+          <SearchPanel
+            updateSearch={onSearchChange}
+            updateQuery={null}
+            hideAdvancedSearch={true}
+            showSearchBar={true}
+            filter={{ search }}
+            disableSearch={serversidePagination.attempt.status === 'processing'}
+          />
+        ),
+      }}
       columns={[
         {
           key: 'name',
           headerText: 'Name',
         },
         {
           altKey: 'options-btn',
-          render: ({ id }) => (
-            <ActionCell id={id} onEdit={onEdit} onDelete={onDelete} />
+          render: (role: RoleResource) => (
+            <ActionCell
+              onEdit={() => onEdit(role.id)}
+              onDelete={() => onDelete(role.id)}
+            />
           ),
         },
       ]}
       emptyText="No Roles Found"
-      pagination={{ pageSize }}
+      pagination={{ pagerPosition: 'top' }}
```

Comment:
idk if i'm tripping... but doesn't this mean we want the 
<img width="1137" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/0faa0a0c-99cf-47e8-b08f-2572ad41fefe">
 ^ paging buttons to be at the top? (it's at the bottom)

Reply:
I actually thought the same exact thing

Reply:
This prop doesn't work for the server side table :(
I started to fix this, but when the pagination buttons are on top, there is quite large spacing around them. I'm afraid that if I fix this, I may cause styling problems somewhere else (where that spacing would be needed). 

I will remove that prop from here :)